core/Enclave/openssl_operation.cpp

```diff
@@ -419,139 +421,298 @@ sgx_status_t rsa_sign(RSA *rsa_prikey,
-        goto out;
-    }
+        if (EVP_PKEY_sign_init(pkey_ctx) <= 0)
+        {
+            log_e("ecall rsa_sign EVP_PKEY_sign_init failed.\n");
+            goto out;
+        }
 
-    // Signature initialization, set digest mode
-    if (EVP_DigestSignInit(mdctx, &pkey_ctx, digestMode, nullptr, evpkey) != 1)
-    {
-        log_e("ecall rsa_sign EVP_DigestSignInit failed.\n");
-        goto out;
-    }
+        if (EVP_PKEY_CTX_set_signature_md(pkey_ctx, digestMode) <= 0)
+        {
+            log_e("ecall rsa_sign EVP_PKEY_CTX_set_signature_md failed.\n");
+            goto out;
+        }
 
-    // set padding mode
-    if (EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, padding_mode) != 1)
-    {
-        log_e("ecall rsa_sign EVP_PKEY_CTX_set_rsa_padding failed.\n");
-        goto out;
-    }
+        // set padding mode
+        if (EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, padding_mode) != 1)
+        {
+            log_e("ecall rsa_sign EVP_PKEY_CTX_set_rsa_padding failed.\n");
+            goto out;
+        }
 
-    if (padding_mode == RSA_PKCS1_PSS_PADDING)
-    {
-        if (EVP_PKEY_CTX_set_rsa_pss_saltlen(pkey_ctx, EVP_MD_size(digestMode)) != 1)
+        if (padding_mode == RSA_PKCS1_PSS_PADDING)
         {
-            log_e("ecall rsa_sign EVP_PKEY_CTX_set_rsa_pss_saltlen failed.\n");
+            if (salten == -1)
+            {
+                if (EVP_PKEY_CTX_set_rsa_pss_saltlen(pkey_ctx, EVP_MD_size(digestMode)) != 1)
+                {
+                    log_e("ecall rsa_sign EVP_PKEY_CTX_set_rsa_pss_saltlen failed.\n");
+                    goto out;
+                }
+            }
+            else
+            {
+                if (EVP_PKEY_CTX_set_rsa_pss_saltlen(pkey_ctx, salten) != 1)
+                {
+                    log_e("ecall rsa_sign EVP_PKEY_CTX_set_rsa_pss_saltlen failed.\n");
+                    goto out;
+                }
+            }
+        }
+
+        if (EVP_PKEY_sign(pkey_ctx, NULL, &temp_signature_size, message, message_len) <= 0)
+        {
+            log_e("ecall ecc_sign EVP_PKEY_sign failed.\n");
             goto out;
         }
-    }
 
-    // update sign
-    if (EVP_DigestUpdate(mdctx, data, data_len) != 1)
-    {
-        log_e("ecall rsa_sign EVP_DigestSignUpdate failed.\n");
-        goto out;
-    }
+        if (EVP_PKEY_sign(pkey_ctx, signature, &temp_signature_size, message, message_len) <= 0)
+        {
+            log_e("ecall ecc_sign EVP_PKEY_sign failed.\n");
+            goto out;
+        }
 
-    // start sign
-    if (EVP_DigestSignFinal(mdctx, NULL, &temp_signature_size) != 1)
-    {
-        log_e("ecall rsa_sign first EVP_DigestSignFinal failed.\n");
-        goto out;
-    }
+        ret = SGX_SUCCESS;
+        break;
+    case EH_MESSAGE_TYPE_RAW:
+        // verify digestmode and padding mode
+        if (padding_mode == RSA_PKCS1_PSS_PADDING)
+        {
+            if (EVP_MD_size(digestMode) * 2 + 2 > (size_t)EVP_PKEY_size(evpkey))
```

Comment:
what does "EVP_MD_size(digestMode) * 2 + 2" mean?

Reply:
https://android.googlesource.com/platform/system/keymaster/+/refs/heads/master/km_openssl/rsa_operation.cpp#264
is it the source? if yes, add it to the comment code.

Reply:
yes, add this url